src/util/rewriter.js

```diff
@@ -0,0 +1,146 @@
+      [target]: (key, value, parents, context) => {
+        const args = {
+          key, value, parents, context
+        };
+        switch (type) {
+          case 'INJECT':
+            ps.forEach(p => Object.assign(value, { [p.target.split('.').pop()]: p.fn(args) }));
+            return value;
+          case 'FILTER':
+            return ps.every(p => p.fn(args));
+          case 'SORT':
+          default:
+            assert(ps.length === 1, 'Only one sort plugin per target allowed!');
+            return ps[0].fn(args);
+        }
+      }
+    }), {});
+};
+
+const compileMeta = (plugins, fields) => {
+  const pluginsByType = pluginTypes.reduce((p, c) => Object.assign(p, { [c]: [] }), {});
+
+  const inactivePlugins = [...plugins];
+  const requiredFields = [...fields];
+  const ignoredFields = new Set();
+
+  for (let i = 0; i < requiredFields.length; i += 1) {
+    const field = requiredFields[i];
+    for (let j = 0; j < inactivePlugins.length; j += 1) {
+      const plugin = inactivePlugins[j];
+      if (
+        field === plugin.target
+        || (plugin.type !== 'INJECT' && field.startsWith(plugin.target))
+      ) {
+        requiredFields.push(...plugin.requires);
+        inactivePlugins.splice(j, 1);
+        j -= 1;
+        pluginsByType[plugin.type].push(plugin);
+        if (plugin.type === 'INJECT' && !plugin.requires.includes(plugin.target)) {
+          ignoredFields.add(plugin.target);
+        }
+      }
+    }
+  }
+
+  return Object.entries(pluginsByType).reduce((p, [type, ps]) => Object.assign(p, {
+    [`${type.toLowerCase()}Cbs`]: compileTargetToCallback(type, ps)
+  }), {
+    toRequest: [...new Set(requiredFields)].filter(e => !ignoredFields.has(e))
+  });
+};
+
+module.exports = (pluginMap) => {
+  const plugins = Object.entries(pluginMap).reduce((prev, [prefix, ps]) => {
+    prev.push(...ps.map(p => p(prefix)));
+    return prev;
+  }, []);
+
+  return (fields) => {
+    assert(Array.isArray(fields));
+
+    const {
+      injectCbs,
+      filterCbs,
+      sortCbs,
+      toRequest
+    } = compileMeta(plugins, fields);
+
+    const injectRewriter = (input, context) => objectScan(Object.keys(injectCbs), {
+      useArraySelector: false,
+      joined: false,
+      filterFn: (key, value, { matchedBy, parents }) => {
+        matchedBy.forEach((m) => {
+          Object.assign(value, injectCbs[m](key, value, parents, context));
+        });
+        return true;
+      }
+    })(input);
+    const filterRewriter = (input, context) => objectScan(Object.keys(filterCbs), {
+      useArraySelector: false,
+      joined: false,
+      filterFn: (key, value, { matchedBy, parents }) => {
+        const result = matchedBy.some(m => filterCbs[m](key, value, parents, context) === true);
+        if (result === false) {
+          const parent = key.length === 1 ? input : parents[0];
+          assert(Array.isArray(parent), 'Should only filter Array entries?');
```

Comment:
I think this was okay as an "iffy" assert in service-search, but now that we're not chasing coverage where this library is used, this can be allowed to be more strict.
```suggestion
          // re-evaluate when use case arises
          assert(Array.isArray(parent), 'Should only filter Array entries.');
```

Reply:
changed it